### PR TITLE
Implement trailing comma for comma-separated lists

### PR DESCRIPTION
**Pull request by mateusrodriguesxyz**
This is a implementation for https://github.com/swiftlang/swift-evolution/pull/2344 gated behind `-enable-experimental-feature TrailingComma`. After feedback from the language steering group I have expanded the original proposal and implementation that I began in https://github.com/apple/swift/pull/71975.

### CODE REVIEW

**Comment by xwu** [comment]
apple/swift-syntax#2689

@swift-ci build toolchain

**Comment by xwu** [comment]
apple/swift-syntax#2689

@swift-ci build toolchain

File: lib/Parse/ParseExpr.cpp
Comment (ahoppen) [inline lib/Parse/ParseExpr.cpp:1918]: Could you check for `Tok.is(tok::r_paren)` instead of comparing the text?

File: lib/Parse/ParseGeneric.cpp
Comment (ahoppen) [inline lib/Parse/ParseGeneric.cpp:409]: Same comment that I have on the swift-syntax PR here: Can we allow trailing comma in inheritance clauses because `associatedtype` declarations don’t have a  clear terminator?

File: lib/Parse/ParseStmt.cpp
Comment (ahoppen) [inline lib/Parse/ParseStmt.cpp:1890]: I think you also need to check for `oper_binary_unspaced` here.

File: lib/Parse/ParseExpr.cpp
Comment (mateusrodriguesxyz) [reply lib/Parse/ParseExpr.cpp:1918]: So... I don't know why but `Tok.is(tok::r_paren)` doesn't work here 🫠

Comment (ahoppen) [reply lib/Parse/ParseExpr.cpp:1918]: Do you know what the token’s kind is? Otherwise I’ll try and start debugging it because I’m curious now.

**Comment by mateusrodriguesxyz** [comment]
> @swift-ci build toolchain

https://ci.swift.org/job/swift-PR-toolchain-macos/1317/

**Comment by xwu** [comment]
@mateusrodriguesxyz Would it be wise to add tests that exercise the code paths where you've made changes in response to @ahoppen?

File: lib/Parse/ParseStmt.cpp
Comment (mateusrodriguesxyz) [reply lib/Parse/ParseStmt.cpp]: @ahoppen do you have a valid example that would `oper_binary_unspaced`? I can't figure out.

**Comment by xwu** [comment]
https://github.com/swiftlang/swift-syntax/pull/2689

@swift-ci build toolchain

File: lib/Parse/ParseStmt.cpp
Comment (ahoppen) [reply lib/Parse/ParseStmt.cpp]: That should be something like the following (no spaces around `!=`). Or am I missing something?

```swift
if true, { (x: () -> Void) in true }!=nil { print(0) }
```

**Comment by xwu** [comment]
Toolchain links:

* [Ubuntu](https://download.swift.org/tmp/pull-request/74522/899/ubuntu2004/PR-ubuntu2004.tar.gz)
* [Windows](https://ci-external.swift.org/job/swift-PR-build-toolchain-windows/5327/artifact/)
* [macOS](https://download.swift.org/tmp/pull-request/74522/1327/xcode/swift-PR-74522-1327-osx.tar.gz)

File: lib/Parse/ParseExpr.cpp
Comment (LucianoPAlmeida) [inline lib/Parse/ParseExpr.cpp:3298]: Nit: Is this line clang-formatted, it looks longer than 80 columns

File: lib/Parse/ParseExpr.cpp
Comment (LucianoPAlmeida) [reply lib/Parse/ParseExpr.cpp:1930]: Also curious, that if we `checkForInputComplete` wouldn't a call to `!isInputIncomplete()` achieve the same as this check `Tok.is(tok::eof) && Tok.getText() == ")"` does? (note condition is different so that is why I think maybe negate it could have same result)

**Comment by mateusrodriguesxyz** [comment]
@hborla as code owner and member of the language steering group do you have advice about how should I proceed now that [SE-0439 has been accepted with modifications](https://forums.swift.org/t/accepted-with-modifications-se-0439-allow-trailing-comma-in-comma-separated-lists/73216/1)? Should I remove all code unrelated with the accepted design or the group wants to keep some of it (e.g: if/guard/while conditions) under an experimental flag? Should I ungate the accepted parts and remove the flag? I would appreciate any direction, thanks!

**Comment by hborla** [comment]
@mateusrodriguesxyz Good question, thanks for asking!

> Should I remove all code unrelated with the accepted design or the group wants to keep some of it (e.g: if/guard/while conditions) under an experimental flag? Should I ungate the accepted parts and remove the flag?

Since this proposal doesn't need the experimental feature flag to be promoted to an upcoming feature flag, I recommend un-gating the accepted parts of the design, and leaving the flag in place to gate the parts of the design that were subsetted out. I also recommend splitting up the tests so that the parts of the tests that need the flag are in a separate file that enable the experimental feature, while the bulk of the tests don't enable the flag.

File: include/swift/Basic/Features.def
Comment (hborla) [inline include/swift/Basic/Features.def:401]: Please switch the argument here to `false`; this controls whether the experimental feature is enable-able in production compiler builds, which this experimental feature shouldn't be. You'll need to `// REQUIRE: asserts` in any tests that use the experimental flag, which is fine!

**Comment by mateusrodriguesxyz** [comment]
@hborla following your recomendations I've ungated the accepted parts and moved the parts of the tests that need the flag  to a diffeent file. I also removed support from `#available` because it really doens't make sense and while running the full suit of parser tests I noticed that it made the diagnostic worse when `*` is missing.

**Comment by xwu** [comment]
@mateusrodriguesxyz Would you mind throwing in tests to ensure that `[Int,]` in type position, `@inline(never,)`, and the like remain invalid? And have you been able to look into which other built-in attributes have custom parameter parsing that doesn't work like a comma-separated parameter list to ensure that we're not adding support for trailing commas in those positions?

**Comment by mateusrodriguesxyz** [comment]
> And have you been able to look into which other built-in attributes have custom parameter parsing that doesn't work like a comma-separated parameter list to ensure that we're not adding support for trailing commas in those positions?

I will look into this as soon as possible.

File: lib/Parse/ParseStmt.cpp
Comment (ahoppen) [inline lib/Parse/ParseStmt.cpp:1528]: Why do we need to check for `Source != AvailabilitySpecSource::Available` here?

File: test/Parse/availability_query_unavailability.swift
Comment (ahoppen) [inline test/Parse/availability_query_unavailability.swift:85]: Would it make sense to keep these deleted tests around to make sure that we now parse them correctly?

**Review by ahoppen — COMMENTED** [review]
Sorry for taking so long to review this. 

I didn’t review the places where trailing comma is now unconditionally enabled vs enabled behind the `TrailingComma` experimental feature because I lost track of which parts of SE-0439 got accepted. I’d like to review that after the proposal has been updated (assuming that it will be).

For the places where trailing commas are currently only allowed behind the experimental feature, could you also add a test case that checks that we reject the syntax if the experimental feature is not passed?

File: lib/Parse/ParseStmt.cpp
Comment (mateusrodriguesxyz) [reply lib/Parse/ParseStmt.cpp:1528]: I did this so `#available` won't support trailing comma. It doesn't really make sense and it leads to worse diagnostic when `*` is missing.

File: test/Parse/availability_query_unavailability.swift
Comment (mateusrodriguesxyz) [reply test/Parse/availability_query_unavailability.swift:85]: I delete here because I added a similar test case to trailing comma test files.

File: lib/Parse/ParseExpr.cpp
Comment (mateusrodriguesxyz) [reply lib/Parse/ParseExpr.cpp:1931]: This is for handling trailing comma in a string interpolation.

**Comment by mateusrodriguesxyz** [comment]
@xwu I've updated the tests with built-in attributes. None of them currently support trailing comma but maybe `backDeployed` and `storageRestrictions` should because they take a variadic list of value as the last argument?

**Comment by mateusrodriguesxyz** [comment]
> For the places where trailing commas are currently only allowed behind the experimental feature, could you also add a test case that checks that we reject the syntax if the experimental feature is not passed?

I have added these tests at the end:
https://github.com/swiftlang/swift/blob/e4ff4846a78633e1361f7f6e735f06ae16771562/test/Parse/trailing-comma.swift#L112-L139

Did you mean this?

File: lib/Parse/ParseExpr.cpp
Comment (mateusrodriguesxyz) [reply lib/Parse/ParseExpr.cpp:3296]: I don't know the `yield` statement but if it's something like `yield(...)` I believe it should. My plan is to go back to the swift-syntax PR to fix any inconsistent once this PR is approved.

File: lib/Parse/ParseExpr.cpp
Comment (ahoppen) [reply lib/Parse/ParseExpr.cpp:1931]: Ah, I see. Could you add that as a comment?

File: lib/Parse/ParseExpr.cpp
Comment (ahoppen) [reply lib/Parse/ParseExpr.cpp:3296]: `yield` can be followed by an expression list without parentheses. Eg. `yield x, y` is valid. But `yield x, ` should be disallowed.

File: lib/Parse/ParseStmt.cpp
Comment (ahoppen) [reply lib/Parse/ParseStmt.cpp:1528]: Why don’t we need to check for `AvailabilitySpecSource::Unavailable`?

File: test/Parse/availability_query_unavailability.swift
Comment (ahoppen) [reply test/Parse/availability_query_unavailability.swift:85]: Ah, I missed that. Thanks.

File: test/Parse/trailing-comma.swift
Comment (ahoppen) [inline test/Parse/trailing-comma.swift:1]: Why do we need to add `-disable-experimental-parser-round-trip`? If we do, that indicates that there is a round-trip bug in the new parser that should be fixed.

**Review by ahoppen — COMMENTED** [review]
Could you also add a test case that checks that we reject the syntax if the experimental feature is not passed?

File: test/Parse/trailing-comma.swift
Comment (mateusrodriguesxyz) [reply test/Parse/trailing-comma.swift:1]: Honestly I just copied this from another test file 😅 I will remove it!

File: lib/Parse/ParseStmt.cpp
Comment (mateusrodriguesxyz) [reply lib/Parse/ParseStmt.cpp:1528]: Makes more sense indeed.

**Comment by mateusrodriguesxyz** [comment]
> Could you also add a test case that checks that we reject the syntax if the experimental feature is not passed?

Sorry, I'm not following. Could you please give an example of what you mean?

**Comment by ahoppen** [comment]
> Sorry, I'm not following. Could you please give an example of what you mean?

I was stupid. Didn’t realize that trailing-comma.swift vs trailing-comman-experimental.swift is exactly what I wanted. Sorry for the confusion.

File: lib/Parse/ParseExpr.cpp
Comment (xwu) [reply lib/Parse/ParseExpr.cpp:3296]: The accepted portions of the proposal are exclusively symmetrically delimited comma-separated lists (i.e., those surrounded by `(...)`, `[...]`, `<...>`, etc.), so this would need to be gated behind a flag.

**Comment by xwu** [comment]
> @xwu I've updated the tests with built-in attributes. None of them currently support trailing comma but maybe backDeployed and storageRestrictions should because they take a variadic list of value as the last argument?

IMO, whatever `@backDeployed` is doing `@available` should also, but as this hasn't been worked through, if all the list handling in built-in attributes are custom and your PR never made any changes to the status quo then we can consider them not part of the proposal and leave it alone.

File: lib/Parse/ParseExpr.cpp
Comment (mateusrodriguesxyz) [reply lib/Parse/ParseExpr.cpp:3296]: @ahoppen do you have an example for this? I tried the following but it doesn't parse::
```
struct Foo {
  var x: Int
  var y: Int {
    _read { yield x, } // ❌ consecutive statements on a line must be separated by ';' | expected expression
  }
}
```

**Comment by mateusrodriguesxyz** [comment]
> if all the list handling in built-in attributes are custom

FWIW `backDeployed` actually uses `parseListItem` with `AllowSepAfterLast` set to `false`.

File: lib/Parse/ParseExpr.cpp
Comment (ahoppen) [reply lib/Parse/ParseExpr.cpp:3296]: Ah, sorry. I did misremember how `yield` works. It does require parentheses if multiple expressions should be yielded, so this is not an issue. Apologies for the confusion.

**Comment by xwu** [comment]
Ship it?

**Comment by mateusrodriguesxyz** [comment]
> Ship it?

Hi @xwu,  just one last thing: Should I remove `@attached` trailing support since it's a built-in attribute?

**Comment by xwu** [comment]
I'm kind of surprised to learn it was included in the first place, as it's not really a list. Or are you referred to the `named(...)` part?

**Comment by mateusrodriguesxyz** [comment]
No, I mean something like `@attached(extension, conformances: P1, P2,)`.

**Review by mikeash — CHANGES_REQUESTED** [review]
There are a lot of formatting changes here. Not sure if we want those changes, but at the very least they should be in their own PR to avoid cluttering the functional changes here, and to make it easier to merge/revert/cherry-pick/etc. this work.

**Comment by xwu** [comment]
Got it. I'm not certain, so the safer route is not to add trailing comma support.

**Comment by mikeash** [comment]
`git-format` is great! Just don't run it on the entire repository. 😄 Clang includes a `clang-format-diff.py` script which makes it easy to run `git-format` on just your changes.

**Comment by mikeash** [comment]
I guess `git-format` got a little too eager? The `clang-format-diff.py` script that comes with clang has worked well for me to apply `clang-format` to just my changes.

**Comment by eeckstein** [comment]
Yes, please do not mass-reformat existing code. It makes it nearly impossible to merge between branches without creating a large amount of conflicts.

**Comment by mateusrodriguesxyz** [comment]
Sorry about the mess. Should I force push to replace the merge commit or make one reverting the formatting?

**Comment by al45tair** [comment]
FWIW, using `git clang-format HEAD^n` where `n` is the number of commits you've made is fine, since that will only reformat things you've changed. Reformatting as you've done here also has the downside that you've pulled in far more reviewers than are likely needed for your actual change.

**Comment by mikeash** [comment]
That's a much nicer looking diff now, thanks!

**Review by ahoppen — APPROVED** [review]

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by evrensen467** [comment]
hey, just a novice here, what does it mean for a feature to be experimental, like this one? i can see in the proposal that this was accepted with modifications, what are the modifications? will we be able to have trailing commas for function arguments?

**Comment by xwu** [comment]
You can refer to the [Swift Forums](https://forums.swift.org) for information about new language features adopted through the Swift Evolution process. In this case, the acceptance with modifications is documented [here](https://forums.swift.org/t/accepted-with-modifications-se-0439-allow-trailing-comma-in-comma-separated-lists/73216).

This feature isn't experimental, and no flags will be required to use it when it ships. It has not, however, shipped in any officially released version of Swift yet. You can check the release notes of any release to see what features are included in that release. You can try out development snapshots which contain new features not yet released by going to the [Swift.org Install](https://www.swift.org/install/) page.

Experimental flags are a way to allow work-in-progress to be integrated and tested; they can be added and removed arbitrarily and, for end users, should be disregarded. User-facing features that are meant to be made available to end users generally go through the [Swift Evolution](https://www.swift.org/swift-evolution/) process, which is conducted on the Swift Forums.

**Comment by ahoppen** [comment]
@mateusrodriguesxyz What’s the state of this feature? I can see that the Swift Evolution proposal [SE-0439](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0439-trailing-comma-lists.md) has been accepted but the features is still guarded behind an experimental feature flag.

**Comment by xwu** [comment]
I believe the experimental feature flag should be guarding those portions of the feature that were not accepted and everything else should be already enabled by default. The proposal text itself has not been edited to incorporate the steering group's decision.

**Comment by mateusrodriguesxyz** [comment]
> The proposal text itself has not been edited to incorporate the steering group's decision.

Yes, the text is just outdated. I've been meaning to update it but I've been pretty busy the last months.
The accepted parts are already ungated and working without experimental flag in the **nightly-main** build.

**Comment by ahoppen** [comment]
Ah, I remember now. Thanks for clarifying, @mateusrodriguesxyz.

**Pull request merged**
